src/sage/libs/gap/element.pyx

```diff
@@ -1508,17 +1510,29 @@ cdef class GapElement_Integer(GapElement):
             sage: huge.sage().ndigits()
             10000
         """
+        cdef UInt* x
+        cdef Int size
+        cdef int c_sign
+        cdef int c_size
+        cdef mpz_t output
         if ring is None:
             ring = ZZ
         if self.is_C_int():
             return ring(GAP_ValueInt(self.value))
         else:
-            # TODO: waste of time!
-            # gap integers are stored as a mp_limb_t and we have a much more direct
-            # conversion implemented in mpz_get_pylong(mpz_srcptr z)
-            # (see sage.libs.gmp.pylong)
-            string = self.String().sage()
-            return ring(string)
+            # gap integers are stored as a mp_limb_t
+            size = GAP_SizeInt(self.value) # count limbs and extract sign
+            if size > 0:
+                c_sign = 1
+                c_size = size
+            elif size < 0:
+                c_sign = -1
+                c_size = -size
+            else: # Something is wrong, fall back to string representation
```

Comment:
do we really need this `else`? As far as I understand from [Int GAP_SizeInt(Obj obj)] (https://github.com/gap-system/gap/blob/4f3a54b2e75c8844277c33f1818285ead7e31df8/src/libgap-api.c#L382) you get `size==0`, the only 3rd possible case, if `self.value==0`.  In any event, there seem to be nothing "wrong" with getting 0 here.

@fingolfin - is this right?

Reply:
In fact the `else` should *never* be triggered: if `self.value == 0` then `self.is_C_int()` should return True, so we don't reach this `if` in the first place.

Reply:
@fingolfin - oops, I should have looked at GAP's src/integer.h

(or is there any other source for docs?)